backend/app/controllers/campaigns_controller.rb

```diff
@@ -48,9 +63,92 @@ def set_campaign
   end
 
   def campaign_params
-    permitted_params = [:name, :description, :promised_amount, :start, :end, :primary_donor_id,
-                        :coupon_denomination, { charity_ids: [] }]
+    top_level_params = %i[charities primary_donor]
+    params.require(top_level_params)
+
+    campaign_params = %i[name description promised_amount start end coupon_denomination]
+    params.require(:campaign).permit(campaign_params)
+  end
+
+  def set_donor
+    primary_donor_params = params[:primary_donor]
+
+    @campaign.primary_donor = PrimaryDonor.find_or_initialize_by(email: primary_donor_params[:email]) do |new_donor|
+      new_donor.name = primary_donor_params[:name]
+    end
+  end
+
+  def create_campaign_charities
+    charity_params = params[:charities]
+
+    @campaign.campaign_charities = charity_params.map do |campaign_charity_param|
+      CampaignCharity.new(charity_id: campaign_charity_param[:charity][:id],
+                          giving_sg_url: campaign_charity_param[:giving_sg_url])
+    end
+  end
+
+  def create_or_update_campaign_charities
```

Comment:
Where does this method update campaign_charities?

Reply:
Done